facilitator/src/logging.rs

```diff
@@ -1,67 +1,169 @@
+impl Display for Severity {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        let string = match self {
+            Severity::Critical => "CRITICAL",
+            Severity::Debug => "DEBUG",
+            Severity::Info => "INFO",
+            Severity::Warning => "WARNING",
+            Severity::Error => "ERROR",
+        };
+        write!(f, "{}", string)
+    }
+}
+
+impl From<Level> for Severity {
+    fn from(slog_level: Level) -> Self {
+        match slog_level {
+            Level::Critical => Severity::Critical,
             Level::Error => Severity::Error,
-            Level::Warn => Severity::Warning,
+            Level::Warning => Severity::Warning,
             Level::Info => Severity::Info,
             Level::Debug => Severity::Debug,
             Level::Trace => Severity::Debug,
         }
     }
 }
 
-#[derive(Debug, Serialize)]
-#[serde(rename_all = "camelCase")]
-struct LogEntry<'a> {
-    message: String,
-    severity: Severity,
-    time: String,
-    module_path: Option<&'a str>,
-    file: Option<&'a str>,
-    line: Option<u32>,
-    thread_id: String,
+/// Options for configuring logging in this application
+pub struct LoggingConfiguration {
+    /// If true, logging output will be forced to JSON format using
+    /// [slog-json][1]. If false, logging format will be determined by detecting
+    /// whether `stderr` is a `tty`. If it is, output is formatted using
+    /// [slog-term][2]. Otherwise, `slog-json` is used.
+    ///
+    /// [1]: https://docs.rs/slog-json
+    /// [2]: https://docs.rs/slog-term
+    pub force_json_output: bool,
+    /// A version string which shall be attached to all log messages
+    pub version_string: &'static str,
+    /// Messages above this log level will be discarded
+    pub log_level: &'static str,
 }
 
-impl<'a> LogEntry<'a> {
-    fn from_record(fmt: &Formatter, record: &'a Record) -> LogEntry<'a> {
-        LogEntry {
-            severity: Severity::from_log_level(record.level()),
-            time: fmt.timestamp().to_string(),
-            message: record.args().to_string(),
-            module_path: record.module_path(),
-            file: record.file(),
-            line: record.line(),
-            thread_id: format!("{:?}", thread::current().id()),
-        }
-    }
+/// Initialize logging resources. On success, returns a root [`slog::Logger`][1]
+/// from which child loggers may be built and a GlobalLoggerGuard which must be
+/// kept live by the caller to enable the `slog_scope` global logger to function
+/// in modules that haven't yet opted into managing their own `slog::Logger`.
+/// Returns an error if `LoggingConfiguration` is invalid.
+///
+/// [1]: https://docs.rs/slog/2.7.0/slog/struct.Logger.html
+pub fn setup_logging(config: &LoggingConfiguration) -> Result<(Logger, GlobalLoggerGuard)> {
+    // We have to box the Drain so that both branches return the same type
+    let drain: Box<dyn Drain<Ok = (), Err = std::io::Error> + Send> =
```

Comment:
Maybe alias the Drain type?

```
type MyDrain = Drain<Ok = (), Err = std::io::Error>;
```

Reply:
[You cannot currently define an alias on traits](https://doc.rust-lang.org/nightly/unstable-book/language-features/trait-alias.html). The best you can do is define a supertrait that captures all the traits you want to alias, but then you have to opt all the types you want to describe with the supertrait in with empty impls. I gave this a try in 4dd69fab579a1665f93e940670d08730bb6fe15d (which I separated from the changes addressing your other comments) and it works, but I'm not sure I like it better given the verbosity.

Reply:
Good point about the problem of trait aliases. I kinda like the version you wound up with, but not enough to stand on principle about it; either way is fine.

Reply:
Well, I had fun figuring it out, and it seems courteous to defer to the reviewer, so I'll the supertrait.